autoprecompiles/src/low_degree_bus_interaction_optimizer.rs

```diff
@@ -0,0 +1,562 @@
+                        .batch_make_range_constraints(replacement.range_constraints)
+                } else {
+                    // Keep the bus interaction as is if a replacement can't be found.
+                    vec![bus_int]
+                }
+            })
+            .collect();
+
+        // Knowing the low-degree functions might help the solver.
+        // The range constraints do not need to be added, because they don't carry information
+        // that is not already implied by the existing bus interactions.
+        self.solver
+            .add_algebraic_constraints(new_constraints.iter().cloned());
+
+        system.algebraic_constraints.extend(new_constraints);
+        system
+    }
+
+    /// Checks whether a bus interaction can be replaced by a low-degree constraint + range checks.
+    /// Returns None if no replacement is found.
+    fn try_replace_bus_interaction(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Option<Replacement<T, V>> {
+        let bus_id = bus_interaction.bus_id.try_to_number()?;
+        if self.bus_interaction_handler.is_stateful(bus_id) {
+            return None;
+        }
+        for input_output_pair in self.possible_input_output_pairs(bus_interaction) {
+            if let Some(low_degree_function) =
+                self.find_low_degree_function(bus_interaction, &input_output_pair)
+            {
+                // Build polynomial constraint
+                let symbolic_inputs = input_output_pair
+                    .inputs
+                    .into_iter()
+                    .map(|input| input.expression)
+                    .collect();
+                let low_degree_function = low_degree_function(symbolic_inputs);
+                let polynomial_constraint =
+                    input_output_pair.output.expression.clone() - low_degree_function;
+
+                // Check degree
+                if polynomial_constraint.degree() > self.degree_bound.identities {
+                    continue;
+                }
+
+                return Some(Replacement {
+                    polynomial_constraint,
+                    range_constraints: self.range_constraints(bus_interaction),
+                });
+            }
+        }
+
+        // No low-degree function found.
+        None
+    }
+
+    /// Given a bus interaction of 2 or 3 unknown fields, finds all combinations of inputs and
+    /// outputs where the input space is small enough.
+    fn possible_input_output_pairs(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Vec<SymbolicInputOutputPair<T, V>> {
+        let unknown_field_indices = bus_interaction
+            .payload
+            .iter()
+            .enumerate()
+            .filter(|(_i, expr)| expr.try_to_number().is_none())
+            .map(|(i, _)| i)
+            .collect::<Vec<_>>();
+        let expressions = unknown_field_indices
```

Comment:
Wouldn't it be easier to compute `expressions` first (and avoid the index access in line 131) and then `unknown_field_indices` is `expressions.keys().collect_vec()`?